In [1]:
import pandas as pd
import numpy as np

import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots

from plotly.offline import iplot, init_notebook_mode
init_notebook_mode(connected=True)

In [2]:
pd.set_option('display.max_columns', None)

In [3]:
df=pd.read_csv('/kaggle/input/gfg-pat/soccer.csv')
df.head()

Country   League   Club       Player Names  Matches_Played  Substitution   \
0   Spain  La Liga  (BET)    Juanmi Callejon              19             16   
1   Spain  La Liga  (BAR)  Antoine Griezmann              36              0   
2   Spain  La Liga  (ATL)        Luis Suarez              34              1   
3   Spain  La Liga  (CAR)       Ruben Castro              32              3   
4   Spain  La Liga  (VAL)      Kevin Gameiro              21             10   

   Mins  Goals     xG  xG Per Avg Match  Shots  OnTarget  Shots Per Avg Match  \
0  1849     11   6.62              0.34     48        20                 2.47   
1  3129     16  11.86              0.36     88        41                 2.67   
2  2940     28  23.21              0.75    120        57                 3.88   
3  2842     13  14.06              0.47    117        42                 3.91   
4  1745     13  10.65              0.58     50        23                 2.72   

   On Target Per Avg Match  Year  
0                     1.03  2016  
1                     1.24  2016  
2                     1.84  2016  
3                     1.40  2016  
4                     1.25  2016

In [4]:
df.shape

(660, 15)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660 entries, 0 to 659
Data columns (total 15 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Country                  660 non-null    object 
 1   League                   660 non-null    object 
 2   Club                     660 non-null    object 
 3   Player Names             660 non-null    object 
 4   Matches_Played           660 non-null    int64  
 5   Substitution             660 non-null    int64  
 6   Mins                     660 non-null    int64  
 7   Goals                    660 non-null    int64  
 8   xG                       660 non-null    float64
 9   xG Per Avg Match         660 non-null    float64
 10  Shots                    660 non-null    int64  
 11  OnTarget                 660 non-null    int64  
 12  Shots Per Avg Match      660 non-null    float64
 13  On Target Per Avg Match  660 non-null    float64
 14  Year                     6

# Q1: What is the total number of goals scored in the entire dataset?




In [6]:
df['Goals'].sum()

7795

The total number of goals scored in the entire dataset is 7795

# Q2: Which player has scored the most goals as a substitution

In [7]:
df.columns = df.columns.str.strip()

In [8]:
df[df['Substitution'] == df['Matches_Played']].nlargest(1, 'Goals')

Country   League  Club Player Names  Matches_Played  Substitution  Mins  \
197   Spain  La Liga  None      Charles              17            17  1861   

     Goals     xG  xG Per Avg Match  Shots  OnTarget  Shots Per Avg Match  \
197     14  13.91              0.71     63        30                 3.22   

     On Target Per Avg Match  Year  
197                     1.53  2018

Charles is the player has scored the most goals as a substitution

# Q3: Find the top 5 players with the most number of goals scored. Also find the total number of goals scored by them

In [9]:
top_5_players_no_of_goals = df.groupby('Player Names')['Goals'].sum().reset_index(name='count').nlargest(5, 'count')

In [10]:
top_5_players_no_of_goals

Player Names  count
259        Lionel Messi    135
93    Cristiano Ronaldo    128
368  Robert Lewandowski    127
86        Ciro Immobile    107
273         Luis Suarez     95

In [11]:
top_5_players_no_of_goals['count'].sum()

592

The total number of goals scored by the top 5 players is 592

# Q4: What are the top 10 countries with the most number of players in the dataset?

In [12]:
top_10_country_no_of_players = df.groupby('Country')['Player Names'].size().reset_index(name='count').nlargest(10, 'count')

In [13]:
top_10_country_no_of_players

Country  count
1        Brazil    100
4       Germany    100
5         Italy    100
7         Spain    100
2       England     80
3        France     60
0   Netherlands     40
6     Portugal      40
8           USA     40

# Q5: # Q1. Create a new dataframe after removing the ourliers for Mins Column in the dataset (conside the upper quantile to be 99 percentile and the lower quantile to be 1 percentile.). After that Find the difference in the number of rows?

* Difference is df.shape(only rows) - df.shape(rows of filtered data only)

In [14]:
new_df = df[(df['Mins'] > df['Mins'].quantile(0.01)) | (df['Mins'] < df['Mins'].quantile(0.99))]

In [15]:
df.shape[0] - new_df.shape[0]

0

After removing the outliers for the column 'Mins' in the dataset, the difference in the number of rows is 0 (nothing)

# Q6: Which top 5 players have scored the maximum number of goals throughout all the years? Display them in descending order

In [16]:
year_wise = df.groupby(['Year', 'Player Names'])['Goals'].sum().reset_index(name='count')

In [17]:
year_wise_top_5_players_no_of_goals = year_wise.groupby('Year').apply(lambda x: x.nlargest(5, 'count')).reset_index(drop=True)

In [18]:
year_wise_top_5_players_no_of_goals

Year               Player Names  count
0   2016          Cristiano Ronaldo     42
1   2016               Lionel Messi     37
2   2016  Pierre-Emerick Aubameyang     31
3   2016         Robert Lewandowski     30
4   2016                 Edin Dzeko     29
5   2017               Lionel Messi     33
6   2017              Ciro Immobile     29
7   2017               Mauro Icardi     29
8   2017         Robert Lewandowski     29
9   2017          Cristiano Ronaldo     26
10  2018               Lionel Messi     36
11  2018       Kylian Mbappe-Lottin     33
12  2018         Fabio Quagliarella     26
13  2018               Duvan Zapata     23
14  2018           Krzysztof Piatek     23
15  2019                Carlos Vela     36
16  2019              Ciro Immobile     36
17  2019         Robert Lewandowski     34
18  2019          Cristiano Ronaldo     31
19  2019         Zlatan Ibrahimovic     31
20  2020            Cyriel Dessers      15
21  2020            Steven Berghuis     15
22  2020            Thiago Galhardo     15
23  2020              Bryan Linssen     14
24  2020                Diego Rossi     14

# Q7. Using Bar Plot show in which year has Cristiano Ronaldo scored the most number of goals?

In [19]:
cr7 = df[df['Player Names'] == 'Cristiano Ronaldo']

In [20]:
x = cr7['Year']
y = cr7['Goals']

fig = go.Figure(data=[go.Bar(x=x, y=y, text=y, textposition='auto')])

fig.update_layout(title='Year-wise Goals by cr7')

iplot(fig)

# Q8. In the year 2019 who has more Goals between Ronaldo and Messi? 
- First print them normally 
- Display the comparison using a bar plot

In [21]:
year19 = df[(df['Year'] == 2019) & ((df['Player Names'] == 'Cristiano Ronaldo') | (df['Player Names'] == 'Lionel Messi'))]

In [22]:
year19

Country   League   Club       Player Names  Matches_Played  Substitution  \
333   Spain  La Liga  (BAR)       Lionel Messi              32             1   
385   Italy  Serie A  (JUV)  Cristiano Ronaldo              33             0   

     Mins  Goals     xG  xG Per Avg Match  Shots  OnTarget  \
333  3067     25  21.63              0.67    159        68   
385  3127     31  27.32              0.83    208        79   

     Shots Per Avg Match  On Target Per Avg Match  Year  
333                 4.93                     2.11  2019  
385                 6.32                     2.40  2019

In [23]:
fig = go.Figure(data=[go.Bar(x=year19['Player Names'], y=year19['Goals'], 
                             text=year19['Goals'], textposition='auto', marker_color=['blue', 'red'])])

fig.update_layout(title='Goals between Ronaldo and Messi', xaxis_title='Player', yaxis_title='Goals')

iplot(fig)

# Q9: Using Pie Chart show the amount of data various Countries are containing in this dataset(use the Country Columns Display the percentage of data as well in the pie chart

In [24]:
countries = df.groupby('Country').size().reset_index(name='count')

In [25]:
fig = px.pie(countries, values='count', names='Country', hover_data=['count'])

fig.update_traces(textposition='inside', textinfo='percent+value', title='Each countries records')

iplot(fig)

# Q10: Display a Heatmap  to show the correlation between the columns in the dataset. The heatmap should have the value of the correlation printed within the boxes.

In [26]:
corr_matrix = df.corr(numeric_only=True)

trace = go.Heatmap(z=corr_matrix.values,
                   x=corr_matrix.columns,
                   y=corr_matrix.columns,
                   colorscale='RdBu',
                   colorbar=dict(title='Correlation'),
                   text=corr_matrix.values.round(2),
                   hoverinfo='text')

layout = go.Layout(title='Correlation Heatmap')

fig = go.Figure(data=[trace], layout=layout)

fig.update_layout(height=800)

iplot(fig)

# hover-over for correlation values

# Q11: Which Country has the maximum median OnTarget Shots among all the Countries in the year 2020?

In [27]:
year2020 = df[df['Year'] == 2020]

In [28]:
year2020.groupby('Country')['OnTarget'].median().reset_index(name='median').nlargest(1, 'median')

Country  median
0   Netherlands    25.5

The country has the maximum median OnTarget Shots among all the Countries in the year 2020 is Netherlands

# Q12. Who has the best average  Goals per minute?
* Consider only those players who have played more than 30 games
* Then Find Goals per minute
* Find Mean of Goals per minute and sort the values

In [29]:
players_filtered = df[df['Matches_Played'] > 30]

players_filtered['goals_per_min'] = players_filtered['Goals'] / players_filtered['Mins']

/tmp/ipykernel_20/1321969036.py:3: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [30]:
players_filtered.groupby('Player Names')['goals_per_min'].mean().reset_index(name='mean').nlargest(1, 'mean')

Player Names      mean
16  Carlos Vela  0.011509

Carlos Vela has the best average Goals per minute

# Q 13. Which Player has the least and most number of missed shots in the dataset?
* Missed Shots=Shots-On Target

In [31]:
df['missed shots'] = df['Shots'] - df['OnTarget']

In [32]:
df.nsmallest(1, 'missed shots')

Country   League   Club    Player Names  Matches_Played  Substitution  \
526   Spain  La Liga  (EIB)  Esteban Burgos               6             1   

     Mins  Goals    xG  xG Per Avg Match  Shots  OnTarget  \
526   589      3  2.85              0.46      5         4   

     Shots Per Avg Match  On Target Per Avg Match  Year  missed shots  
526                 0.81                     0.65  2020             1

In [33]:
df.nlargest(1, 'missed shots')

Country   League   Club       Player Names  Matches_Played  Substitution  \
385   Italy  Serie A  (JUV)  Cristiano Ronaldo              33             0   

     Mins  Goals     xG  xG Per Avg Match  Shots  OnTarget  \
385  3127     31  27.32              0.83    208        79   

     Shots Per Avg Match  On Target Per Avg Match  Year  missed shots  
385                 6.32                      2.4  2019           129

Player has the least and most number of missed shots in the dataset are Esteban Burgos and Cristiano Ronaldo respectively.

# Q14: How many countries have had players who played more than 35 matches? Create a bar plot to display the count for each country.

In [34]:
filtered_players = df[df['Matches_Played'] > 35]

filtered_players = filtered_players.groupby('Country')['Player Names'].size().reset_index(name='count')

In [35]:
x = filtered_players['Country']
y = filtered_players['count']

fig = go.Figure(data=[go.Bar(x=x, y=y, text=y, textposition='auto')])

fig.update_layout(title='Countries with Players Having More Than 35 Matches')

iplot(fig)

# Q15: Draw a Line Chart to show the comparison of the top 5 highest scoring players in the dataset throughout the years(2016-2020)

In [36]:
player_goals = df.groupby(['Player Names', 'Year'])['Goals'].sum().reset_index()

top_players = player_goals.groupby('Player Names')['Goals'].sum().nlargest(5).index

In [37]:
fig = go.Figure()

for player in top_players:
    player_data = player_goals[player_goals['Player Names'] == player]
    fig.add_trace(go.Scatter(x=player_data['Year'].astype(str), y=player_data['Goals'], mode='lines', name=player, line_shape='spline'))

fig.update_layout(title='Top 5 Highest Scoring Players (2016-2020)',
                  xaxis_title='Year',
                  yaxis_title='Goals')

iplot(fig)